apps/prairielearn/src/lib/bot-grading.ts

```diff
@@ -0,0 +1,182 @@
+import * as cheerio from 'cheerio';
+import { OpenAI } from 'openai';
+import { z } from 'zod';
+
+import * as error from '@prairielearn/error';
+import { loadSqlEquiv, queryRow, queryRows } from '@prairielearn/postgres';
+
+import { config } from '../lib/config.js';
+import {
+  InstanceQuestionSchema,
+  SubmissionSchema,
+  VariantSchema,
+  Question,
+  Course,
+  AssessmentQuestion,
+} from '../lib/db-types.js';
+import * as manualGrading from '../lib/manualGrading.js';
+import { buildQuestionUrls } from '../lib/question-render.js';
+import { getQuestionCourse } from '../lib/question-variant.js';
+import { createServerJob } from '../lib/server-jobs.js';
+import * as questionServers from '../question-servers/index.js';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const SubmissionVariantSchema = z.object({
+  variant: VariantSchema,
+  submission: SubmissionSchema,
+});
+
+export async function botGrade({
+  course,
+  course_instance_id,
+  question,
+  assessment_question,
+  urlPrefix,
+}: {
+  question: Question;
+  course: Course;
+  course_instance_id?: string;
+  assessment_question: AssessmentQuestion;
+  urlPrefix: string;
+}): Promise<string> {
+  // if OpenAI API Key and Organization are not provided, throw error
+  if (!config.openAiApiKey || !config.openAiOrganization) {
+    throw new error.HttpStatusError(501, 'Not implemented (feature not available)');
+  }
+  const openaiconfig = { apiKey: config.openAiApiKey, organization: config.openAiOrganization };
+
+  const question_course = await getQuestionCourse(question, course);
+
+  const serverJob = await createServerJob({
+    courseId: course.id,
+    courseInstanceId: course_instance_id,
+    assessmentId: assessment_question.assessment_id,
+    authnUserId: '1',
```

Comment:
Actually Seth and I were thinking earlier that we might want to create a new user just for bot grading so that it can show the work as being graded using the bot grade function? For now can I just change it to the user making the request.